NETOBSERV-1323: Secondary interface support blog (#56)
netobserv/documents: blogs/secondary_interface/secondary_interface.md

   Employ the netlink package to register when network interfaces are created or deleted
   within each network namespace. This will enable your eBPF agent to monitor interface
   changes on a per-namespace basis.

3. Attach/Detach eBPF TC Hooks:
   Add support to your eBPF agent to attach and detach eBPF Traffic Control (TC) hooks
   for network interfaces in non-default network namespaces. This step is crucial for
   monitoring and controlling network traffic within these namespaces.

Here is a sample configuration to build and test this topology

- Install the SRIOV operator in the environment.
- Identify the SRIOV capable device on the node.
- Label the node which has SRIOV interface with the label
  `feature.node.kubernetes.io/network-sriov.capable=true`
- Create the SriovNetworkNodePolicy object

```yaml
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetworkNodePolicy
metadata:
  name: mypolicy
  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  nodeSelector:
    feature.node.kubernetes.io/network-sriov.capable: "true"
  priority: 99
  numVfs: 50
  nicSelector:
    pfNames: ["ens7f0np0#25-49"]
  deviceType: netdevice
```

> - Create the SriovNetwork object

Reviewer: ```suggestion
- Create the SriovNetwork object. This will create net-attach-def in `openshift-sriov-network-operator` namespace
```

```yaml
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetwork
metadata:
  name: sriov-test
  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  networkNamespace: test
  ipam: '{ "type": "static", "addresses": [{"address": "192.168.122.71/24"}]}'
```

> - Create a test pod for testing note sriovNetwork object will trigger a

Reviewer: ```suggestion
- Create a test pod using `SRIOVNetwork` created above denoted by annotation `k8s.v1.cni.cncf.io/networks: "sriov-test"`
```

  net-attach-def in the test namespace.

```yaml
apiVersion: v1
kind: Pod
metadata:

>   name: httpd-2

Reviewer: ```suggestion
  name: httpd-2
  namespace: openshift-sriov-network-operator
```

  labels:
    app: sriov
  annotations:
    k8s.v1.cni.cncf.io/networks: "sriov-test"
spec:
  containers:
  - name: httpd
    command: ["sleep", "30d"]
    image: registry.redhat.io/rhel8/support-tools
    ports:
    - containerPort: 8080
    securityContext:
      allowPrivilegeEscalation: false
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop:
        - ALL
```

- Deploy network observability operator

- Create `FollowCollector` object with `privileged` set

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true
```

> - netobservability flows filter on Pod VF interface `net1`

Reviewer: ```suggestion
- Network observability flows filter on Pod VF interface `net1`
```

# Network Observability secondary interface support

By: Mohamed S. Mahmoud

Network observability for secondary interfaces with Multus and SR-IOV plugins
in Kubernetes can be a complex task, but it's crucial for monitoring and
troubleshooting network issues in a Kubernetes cluster.

> Here's an overview of how you can achieve network observability for secondary

Reviewer 2: I'd add something here about how we've already added this ability with network observability operator, in OCP, with our 1.4 version

Author: u are referring to this section https://github.com/netobserv/documents/pull/56/files#diff-a12dd20328c486e3729e49cbcc9d666fb2371b876df01772d54dc35b2ae31b29R26

  - name: httpd
    command: ["sleep", "30d"]
    image: registry.redhat.io/rhel8/support-tools
    ports:
    - containerPort: 8080
    securityContext:
      allowPrivilegeEscalation: false
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop:
        - ALL
```

- Deploy network observability operator

- Create `FollowCollector` object with `privileged` set

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true
```

- Network observability raw flows filtered on Pod VF interface `net1`

```json
{
  "AgentIP": "10.73.116.62",
  "Bytes": 406,
  "DstAddr": "192.168.122.72",
  "DstMac": "06:D8:93:DB:56:9A",
  "DstPort": 60256,
  "Duplicate": false,
  "Etype": 2048,
  "Flags": 784,
  "FlowDirection": "1",
  "IfDirection": 1,
  "Interface": "net1",
  "K8S_ClusterName": "5846f979-9d5b-4a1b-b08d-aee52e6b58a2",
  "Packets": 4,
  "Proto": 6,
  "SrcAddr": "192.168.122.71",
  "SrcMac": "AE:39:F4:7E:03:FE",
  "SrcPort": 8080,
  "TimeFlowEndMs": 1694120614010,
  "TimeFlowStartMs": 1694120614009,
  "TimeReceived": 1694120618,
  "app": "netobserv-flowcollector"
}

> ```

Reviewer 2: At the end, here, I'd reiterate that you can do all the above, or you can just install network observability operator 1.4 or higher and have it all done for you -- and maybe point to our repositories or something else -- I think there's a great opportunity here to promote the work you've done in the context of network obs. and maybe get more people to try

Author: hmm I think there is misunderstanding , we still need to provision SRIOV, SRIOV policy etc that hasn't changed what is new here is with turning on the operator know and the new work done in the agent in 1.4 we know can see those flows

Author: I will add the repo and feedback sections like in another blogs

# Network Observability secondary interface support

By: Mohamed S. Mahmoud

Network observability for secondary interfaces with Multus and SR-IOV plugins
in Kubernetes can be a complex task, but it's crucial for monitoring and
troubleshooting network issues in a Kubernetes cluster.

## Overview of how you can achieve network observability for secondary interface

<p align="center">
  <img src="images/sriov.png" alt="logo" width="75%"/>
</p>

> 1. **Multus CNI Plugin**: Multus is a CNI (Container Network Interface) plugin

Reviewer 3: might be good to add some links to doc and/or projects page, e.g https://docs.openshift.com/container-platform/latest/networking/multiple_networks/understanding-multiple-networks.html or https://github.com/openshift/multus-cni
and same for sriov
?

# Network Observability secondary interface support

By: Mohamed S. Mahmoud

Network observability for secondary interfaces with Multus and SR-IOV plugins
in Kubernetes can be a complex task, but it's crucial for monitoring and
troubleshooting network issues in a Kubernetes cluster.

## Overview of how you can achieve network observability for secondary interface

<p align="center">
  <img src="images/sriov.png" alt="logo" width="75%"/>
</p>

1. **Multus CNI Plugin**: Multus is a CNI (Container Network Interface) plugin
    for Kubernetes that allows you to attach multiple network interfaces to pods.
    It's commonly used when you want to use SR-IOV for attaching secondary
    interfaces to your pods.

2. **SR-IOV Plugin**: SR-IOV (Single Root I/O Virtualization) is a technology that
   enables the partitioning of a single PCIe network adapter into multiple virtual
   functions (VFs).
   Pods can then use these VFs as secondary network interfaces, achieving higher
   network performance and isolation.

> ## Network observability eBPF agent support for the secondary interface

Author: this section was my attempt to show the required code changes to the agent added in 1.4 to support this functionality

> ## Network observability eBPF agent support for the secondary interface

Reviewer 4: Ok, in that case, I misunderstood. You're saying that these are the changes you made to the code to make this work. Discussion via Slack about alt. titles here..

2. Using netlink Package:
   Employ the netlink package to register when network interfaces are created or deleted
   within each network namespace. This will enable your eBPF agent to monitor interface
   changes on a per-namespace basis.

3. Attach/Detach eBPF TC Hooks:
   Add support to your eBPF agent to attach and detach eBPF Traffic Control (TC) hooks
   for network interfaces in non-default network namespaces. This step is crucial for
   monitoring and controlling network traffic within these namespaces.

## SRIOV objects configuration example

1. Install the SRIOV operator in the environment.
2. Identify the SRIOV capable device on the node.
3. Label the node which has SRIOV interface with the label
  `feature.node.kubernetes.io/network-sriov.capable=true`
4. Create the SriovNetworkNodePolicy object

```yaml
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetworkNodePolicy
metadata:
  name: mypolicy
  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  nodeSelector:
    feature.node.kubernetes.io/network-sriov.capable: "true"
  priority: 99
  numVfs: 50
  nicSelector:
    pfNames: ["ens7f0np0#25-49"]
  deviceType: netdevice
```

> 5. Create the SriovNetwork object. This will create net-attach-def in
>   `openshift-sriov-network-operator` namespace

Reviewer 4: ```suggestion
5. Create the `SriovNetwork` object. This will create net-attach-def in the
  `openshift-sriov-network-operator` namespace.
```

```yaml
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetwork
metadata:
  name: sriov-test
  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  networkNamespace: test
  ipam: '{ "type": "static", "addresses": [{"address": "192.168.122.71/24"}]}'
```

> 6. Create a test pod using `SRIOVNetwork` created above denoted by annotation

Reviewer 4: ```suggestion
6. Create a test pod using the `SRIOVNetwork` object created above and denoted by annotation.
```

  `k8s.v1.cni.cncf.io/networks: "sriov-test"`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: httpd-2
  namespace: openshift-sriov-network-operator
  labels:
    app: sriov
  annotations:
    k8s.v1.cni.cncf.io/networks: "sriov-test"
spec:
  containers:
  - name: httpd
    command: ["sleep", "30d"]
    image: registry.redhat.io/rhel8/support-tools
    ports:
    - containerPort: 8080
    securityContext:
      allowPrivilegeEscalation: false
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop:
        - ALL
```

> ## Network observerability configuration

Reviewer 4: ```suggestion
## Configuring the network observerability operator to work with SR-IOV
```

Reviewer 4: Would this work?

> 1. Deploy network observability operator

Reviewer 4: ```suggestion
1. Deploy the network observability operator.
```

> 2. Create `FollowCollector` object with `privileged` set to `true`

Reviewer 4: ```suggestion
2. Create the `FollowCollector` object with `privileged` set to `true`.
```

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true
```

## Sample Network observability raw flow output by filtering on Pod VF interface `net1`

```json
{
  "AgentIP": "10.73.116.62",
  "Bytes": 406,
  "DstAddr": "192.168.122.72",
  "DstMac": "06:D8:93:DB:56:9A",
  "DstPort": 60256,
  "Duplicate": false,
  "Etype": 2048,
  "Flags": 784,
  "FlowDirection": "1",
  "IfDirection": 1,
  "Interface": "net1",
  "K8S_ClusterName": "5846f979-9d5b-4a1b-b08d-aee52e6b58a2",
  "Packets": 4,
  "Proto": 6,
  "SrcAddr": "192.168.122.71",
  "SrcMac": "AE:39:F4:7E:03:FE",
  "SrcPort": 8080,
  "TimeFlowEndMs": 1694120614010,
  "TimeFlowStartMs": 1694120614009,
  "TimeReceived": 1694120618,
  "app": "netobserv-flowcollector"
}
```

## Feedback

> We hope you liked this article !

Reviewer 5: remove this line. Otherwise LGTM

2. Using netlink Package:
   Employ the netlink package to register when the network interfaces are created or deleted
   within each network namespace. This will enable the eBPF agent to monitor the interface
   changes on a per-namespace basis.

3. Attach/Detach eBPF TC Hooks:
   Add support to the eBPF agent to attach and detach eBPF Traffic Control (TC) hook
   for network interfaces in non-default network namespaces. This step is crucial for
   monitoring and controlling network traffic within these network namespaces.

## Configuring SR-IOV objects

1. Install the SR-IOV operator in the environment.
2. Identify the SR-IOV capable device on the node.
3. Label the node which has SR-IOV interface with the label
  `feature.node.kubernetes.io/network-sriov.capable=true`
4. Create the `SriovNetworkNodePolicy` object

```yaml
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetworkNodePolicy
metadata:
  name: mypolicy
  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  nodeSelector:
    feature.node.kubernetes.io/network-sriov.capable: "true"
  priority: 99
  numVfs: 50
  nicSelector:
    pfNames: ["ens7f0np0#25-49"]

>   deviceType: netdevice

Reviewer 6: Should we go in details for each of these fields as we do in OCP doc with bullet points <1> <2> ... ?

Author: I don't think so as the scope of this blog is to show how to use netobserv for sriov multus i added the ocp doc that has all details and context

  namespace: openshift-sriov-network-operator
spec:
  resourceName: netdeviceresource
  networkNamespace: test
  ipam: '{ "type": "static", "addresses": [{"address": "192.168.122.71/24"}]}'
```

6. Create a test pod using the `SRIOVNetwork` object created above and
   denoted by annotation, `k8s.v1.cni.cncf.io/networks: "sriov-test"`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: httpd-2
  namespace: openshift-sriov-network-operator
  labels:
    app: sriov
  annotations:
    k8s.v1.cni.cncf.io/networks: "sriov-test"
spec:
  containers:
  - name: httpd
    command: ["sleep", "30d"]
    image: registry.redhat.io/rhel8/support-tools
    ports:
    - containerPort: 8080
    securityContext:
      allowPrivilegeEscalation: false
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop:
        - ALL
```

## Configuring the network observability operator to work with SR-IOV

1. Deploy the network observability operator.

2. Create the `FollowCollector` object with `privileged` set to `true`.

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true
```

> 

Reviewer 3: Maybe add a sentence here for transitioning ? Just telling what is going to happen then: netobserv deploys its components, the agents discover all the interfaces and attach their probes, flows start being collected ...

> ## Sample Network observability raw flow output by filtering on Pod VF interface `net1`

Reviewer 3: Maybe guide a little bit the user on how they get that: e.g. by opening the console plugin and looking in the Traffic Flows table (btw we could add a screenshot here, with a flow captured from interface net1)

    ebpf:
      privileged: true
```

Network observability operator will deploy it's components (eBPF agent, flowlogs pipeline and
console plugin), the eBPF agent will start discovering all the interfaces and attach the eBPF
hooks, flows start being collected

## Sample Network observability raw flow output by filtering on Pod VF interface `net1`

By opening the console plugin and looking in the Traffic Flows table and filter by
`Network interface name == net` will get a raw flow like the following for ICMP ping
packets as an example

```json
{
  "AgentIP": "10.73.116.62",
  "Bytes": 490,
  "DstAddr": "192.168.122.71",
  "DstMac": "AE:39:F4:7E:03:FE",
  "Duplicate": false,
  "Etype": 2048,
  "FlowDirection": "0",
  "IcmpCode": 0,
  "IcmpType": 0,
  "IfDirection": 0,
  "Interface": "net1",
  "K8S_ClusterName": "5846f979-9d5b-4a1b-b08d-aee52e6b58a2",
  "Packets": 5,
  "Proto": 1,
  "SrcAddr": "192.168.122.72",
  "SrcMac": "06:D8:93:DB:56:9A",
  "TimeFlowEndMs": 1694117793468,
  "TimeFlowStartMs": 1694117789372,
  "TimeReceived": 1694117793,
  "app": "netobserv-flowcollector"
}
```

<p align="center">
  <img src="images/sriov_console.png" alt="logo" width="75%"/>
</p>

> Todo: working on flow enrichment to replace those `n/a` with the proper kubernetes object string

Reviewer 3: yeah .. any idea why this isn't enriched? Those IPs are not recognized in FLP ; ie. they are not referenced in the Pod specs/status ?
If so, perhaps at the moment you could take another screenshot with different columns, because it doesn't look very nice here; ie. set columns to Timestamps, source/dest IPs, Interface, bytes/pkt counters .. and whatever else you find useful

Author: This was icmp do no ports also we don’t have the setup anymore I was expecting pod names but I am not sure we have an rfe created to follow up https://issues.redhat.com/browse/NETOBSERV-1289

Author: I think for now I will remove the picture and the todo note till we have a better test and img

Author: added new pictures for TCP with populated fields
